cep-16.md

```diff
@@ -0,0 +1,260 @@
+
+### Repodata shard
+
+Individual shards are stored under the URL `<shard_base_url>/<subdir>/shards/<sha256>.msgpack.zst`. Where the `sha256` is the lower-case hex representation of the bytes from the index. It is a zstandard compressed msgpack file that contains the metadata of the package. The `<shard_base_url>` is defined in [Authentication](#authentication).
+
+The files are content-addressable which makes them ideal to be served through a CDN. They SHOULD be served with `Cache-Control: immutable` header.
+
+The shard contains the repodata information that would otherwise have been found in the `repodata.json` file. 
+It is a dictionary that contains the following keys:
+
+**Example (written in JSON for readability):**
+
+```js
+{
+  // dictionary of .tar.bz2 files
+  "packages": {
+    "rich-10.15.2-pyhd8ed1ab_1.tar.bz2": {
+      "build": "pyhd8ed1ab_1",
+      "build_number": 1,
+      "depends": [
+        "colorama >=0.4.0,<0.5.0",
+        "commonmark >=0.9.0,<0.10.0",
+        "dataclasses >=0.7,<0.9",
+        "pygments >=2.6.0,<3.0.0",
+        "python >=3.6.2",
+        "typing_extensions >=3.7.4,<5.0.0"
+      ],
+      "license": "MIT",
+      "license_family": "MIT",
+      "md5": "2456071b5d040cba000f72ced5c72032",
+      "name": "rich",
+      "noarch": "python",
+      "sha256": "a38347390191fd3e60b17204f2f6470a013ec8753e1c2e8c9a892683f59c3e40",
+      "size": 153963,
+      "subdir": "noarch",
+      "timestamp": 1638891318904,
+      "version": "10.15.2"
+    }
+  },
+ // dictionary of .conda files
+  "packages.conda": {
+    "rich-13.7.1-pyhd8ed1ab_0.conda": {
+      "build": "pyhd8ed1ab_0",
+      "build_number": 0,
+      "depends": [
+        "markdown-it-py >=2.2.0",
+        "pygments >=2.13.0,<3.0.0",
+        "python >=3.7.0",
+        "typing_extensions >=4.0.0,<5.0.0"
+      ],
+      "license": "MIT",
+      "license_family": "MIT",
+      "md5": "ba445bf767ae6f0d959ff2b40c20912b",
+      "name": "rich",
+      "noarch": "python",
+      "sha256": "2b26d58aa59e46f933c3126367348651b0dab6e0bf88014e857415bb184a4667",
+      "size": 184347,
+      "subdir": "noarch",
+      "timestamp": 1709150578093,
+      "version": "13.7.1"
+    }
+  },
+  // list of strings of keys (filenames) that were removed from either packages or packages.conda
+  "removed": [
+    "rich-10.15.1-pyhd8ed1ab_1.tar.bz2"
+  ]
+}
+```
+
+The `sha256` and `md5` from the original repodata fields are converted from their hex representation to bytes. 
+This is done to reduce the overall file size of the shards.
+
+Implementers SHOULD ignore unknown keys, this allows adding additional keys to the format in the future without breaking old versions of tools.
+
+Although these files can become relatively large (100s of kilobytes) typically for a large case (conda-forge) these files remaing very small, e.g. 100s of bytes to a couple of kilobytes.
```

Comment:
❤️

Reply:
What if we allow a shard to contain the metadata of more than one package, so that multiple package names could point to the same shard? It is a nuisance to download many 300-byte shards.

Reply:
Yeah we could, but it comes with some downsides:

- cache might not be as efficient as multiple packages might change at different frequencies and cache becomes invalid more often
- it's more complicated on the implementation as packages don't map 1-1 to shards, but that would not be too bad.

We could keep the current repodata_shards index structure and use the same hash for multiple package names (that would probably compress decently). 

However, I am not sure if the tradeoff is worth it. 

Also it will start to look a bit like `zchunk` then :)

Reply:
I tried [using a buffer to combine sequential (ordered by package name) shards](https://github.com/dholth/conda-test-data/commit/d5e4c74aae07125055640a537473e8116fd4454a) when the combined size is not > 8192 bytes compressed. So we still wind up with plenty of ~300 byte single package shards if it sees (large shard, small shard, large shard) but it's all in order.

That scheme gives 7987 files and 10883 package names on a conda-forge/linux-64 snapshot, probably taking 75% of the time to download if we happen to want to download everything. Unlike zchunk we have the option of not maintaining a complete local copy of the data.

The most packages combined into a single shard are these fifteen.

```
['perl-bit-vector-7.4-pl5321h166bdaf_0.tar.bz2',
 'perl-capture-tiny-0.48-pl5321ha770c72_1.tar.bz2',
 'perl-carp-assert-0.21-pl526_0.tar.bz2',
 'perl-class-method-modifiers-2.13-pl5321ha770c72_0.tar.bz2',
 'perl-clone-0.46-pl5321h166bdaf_0.tar.bz2',
 'perl-compress-raw-bzip2-2.201-pl5321h166bdaf_0.tar.bz2',
 'perl-compress-raw-zlib-2.202-pl5321h166bdaf_0.tar.bz2',
 'perl-data-dumper-2.183-pl5321h166bdaf_0.tar.bz2',
 'perl-b-hooks-endofscope-0.26-pl5321ha770c72_0.conda',
 'perl-b-hooks-endofscope-0.28-pl5321ha770c72_0.conda',
 'perl-class-data-inheritable-0.09-pl5321ha770c72_0.conda',
 'perl-class-load-0.25-pl5321ha770c72_0.conda',
 'perl-class-load-xs-0.10-pl5321h0b41bf4_0.conda',
 'perl-class-method-modifiers-2.14-pl5321ha770c72_0.conda',
 'perl-class-method-modifiers-2.15-pl5321ha770c72_0.conda',
 'perl-cpan-meta-check-0.014-pl5321ha770c72_0.conda',
 'perl-data-dumper-2.183-pl5321hd590300_0.conda',
 'perl-data-dumper-concise-2.023-pl5321ha770c72_0.conda',
 'perl-data-optlist-0.112-pl5321ha770c72_0.conda',
 'perl-data-optlist-0.113-pl5321ha770c72_0.conda',
 'perl-data-optlist-0.114-pl5321ha770c72_0.conda']
```

Reply:
I am wondering, if you need everything, wouldn't you rather download the `repodata.json` file instead? I think for us that wasn't the use case we optimized for.

Reply:
I'm imagining a situation in which there is only repodata_shards. Suppose we are in the middle of processing a bunch of queries on the SubdirData and suddenly we encounter a query for "any package name, md5=x". Are we more likely to reject the query, download everything, or eject to downloading an alternate format?

Reply:
Why not deprecate it, and grab the repodata.json in the meantime? Tell people to use proper lockfiles? :)

Reply:
What if we simplified the implementation, and included one repodata shard with all the packages in it, and an index that pointed to that singular shard 10k times 🤣